### DISCUSSION TITLE
Using @nestjs/schedule module

### DISCUSSION THREAD

**Discussion by lupeaserban** [vendurehq/vendure#2464] [Q&A] [answered]
Hello,

Has anyone implemented cron jobs using the [@nestjs/schedule](https://docs.nestjs.com/techniques/task-scheduling) module? 

I've created a plugin in which a provider uses the @Cron decorator. I have Vendure running with 2 separate workers and the problem is both workers and main app are registering the cron therefore triggering the same action 3 times. A way I found to handle this was to add the plugin when bootstraping one of the workers, and not add it to the main vendure-config file.

`const workerConfig = mergeConfig(config, {
      plugins: [CronPlugin],
  });
`
`bootstrapWorker(workerConfig)`

Is this good practice? (I guess I'm overwriting the plugins field in `vendure-config.ts`) 

What if i want to register the plugin only on the main app and not having it picked up by the workers?

Also, when working on this I noticed that all plugins are registered multiple times by the workers and main app. Is this normal?

**Comment by michaelbromley** [comment, marked as answer]
Hi,

Yes, just using the Cron decorator is not suitable for multiple instances. I discuss that in [this thread](https://github.com/vendure-ecommerce/vendure/issues/1425#issuecomment-1049922201), and long-term I would like to support scheduled tasks in a more native way.

Your solution is fine, although the way you do it there will _replace_ all the plugins with just the CronPlugin, which might not be what you want. From the [docs](https://docs.vendure.io/reference/typescript-api/configuration/merge-config/):

> Arrays do not get merged, they are treated as a single value that will be replaced. So if merging the plugins array, you must explicitly concatenate the array.

So you'd want it like:

```ts
const workerConfig = mergeConfig(config, { plugins: [...config.plugins, CronPlugin], }); 
```

> What if i want to register the plugin only on the main app and not having it picked up by the workers?

Inject the [ProcessContext](https://docs.vendure.io/reference/typescript-api/common/process-context/) into your class and use it to check whether it is being run in the server or worker context.

> Also, when working on this I noticed that all plugins are registered multiple times by the workers and main app. Is this normal?

A plugin should be registered once for each instance (worker or server). If you are seeing otherwise and can reproduce it, open an issue please.

**Reply by lupeaserban** [reply]
Thank you
